src/main/scala/org/broadinstitute/hail/utils/TextTableReader.scala

```diff
@@ -0,0 +1,198 @@
+import org.broadinstitute.hail.utils.StringEscapeUtils._
+
+object TextTableConfiguration {
+  final val defaultTypes = Map.empty[String, Type]
+  final val defaultSep = "\t"
+  final val defaultMissing = "NA"
+  final val defaultComment: Option[String] = None
+  final val defaultNoHeader: Boolean = false
+  final val defaultNoImpute: Boolean = false
+}
+
+case class TextTableConfiguration(
+  types: Map[String, Type] = TextTableConfiguration.defaultTypes,
+  commentChar: Option[String] = TextTableConfiguration.defaultComment,
+  separator: String = TextTableConfiguration.defaultSep,
+  missing: String = TextTableConfiguration.defaultMissing,
+  noHeader: Boolean = TextTableConfiguration.defaultNoHeader,
+  noImpute: Boolean = TextTableConfiguration.defaultNoImpute)
+
+object TextTableReader {
+
+  val booleanRegex = """^([Tt]rue)|([Ff]alse)|(TRUE)|(FALSE)$"""
+  val intRegex = """^-?\d+$"""
+  val doubleRegex = """^[-+]?[0-9]*\.?[0-9]+([eE][-+]?[0-9]+)?$"""
+  val variantRegex = """^.+:\d+:[ATGC]+:([ATGC]+|\*)(,([ATGC]+|\*))*$"""
+  val headToTake = 20
+
+  def guessType(values: Seq[String], missing: String): Option[Type] = {
+    require(values.nonEmpty)
+
+    val size = values.size
+    val booleanMatch = values.exists(value => value.matches(booleanRegex))
+    val allBoolean = values.forall(value => value.matches(booleanRegex) || (value == missing))
+
+    val variantMatch = values.exists(value => value.matches(variantRegex))
+    val allVariant = values.forall(value => value.matches(variantRegex) || (value == missing))
+
+    val doubleMatch = values.exists(value => value.matches(doubleRegex))
+    val allDouble = values.forall(value => value.matches(doubleRegex) || (value == missing))
+
+    val intMatch = values.exists(value => value.matches(intRegex))
+    val allInt = values.forall(value => value.matches(intRegex) || (value == missing))
+
+    if (values.forall(_ == missing))
+      None
+    else if (allBoolean && booleanMatch)
+      Some(TBoolean)
+    else if (allVariant && variantMatch)
+      Some(TVariant)
+    else if (allInt && intMatch)
+      Some(TInt)
+    else if (allDouble && doubleMatch)
+      Some(TDouble)
+    else
+      Some(TString)
+  }
+
+  def read(sc: SparkContext,
+    files: Array[String],
+    config: TextTableConfiguration): (TStruct, RDD[WithContext[Annotation]]) = read(sc, files, config.types,
+    config.commentChar, config.separator, config.missing, config.noHeader, config.noImpute)
+
+  def read(sc: SparkContext,
+    files: Array[String],
+    types: Map[String, Type] = TextTableConfiguration.defaultTypes,
+    commentChar: Option[String] = TextTableConfiguration.defaultComment,
+    separator: String = TextTableConfiguration.defaultSep,
+    missing: String = TextTableConfiguration.defaultMissing,
+    noHeader: Boolean = TextTableConfiguration.defaultNoHeader,
+    noImpute: Boolean = TextTableConfiguration.defaultNoImpute): (TStruct, RDD[WithContext[Annotation]]) = {
```

Comment:
I don't like this duplication.  Is `TextTableReader.read(sc, files, TextTableConfiguration(param = override))` so bad?  Moreover, this function is never used, except with all default arguments, and you can make the other version have a default argument for the config.